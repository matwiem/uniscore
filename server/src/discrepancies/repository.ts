import { Operation } from 'fast-json-patch'
import { InvalidFilterError } from '@src/discrepancies/errors'

export type NodeID = string

export const ALL_DISCREPANCY_SUBJECTS = ['GAME', 'TEAM', 'PLAYER'] as const
type DiscrepancySubjectTuple = typeof ALL_DISCREPANCY_SUBJECTS
export type DiscrepancySubject = DiscrepancySubjectTuple[number]

export type EventMeta = {
    gameId: NodeID
    homeTeamId: NodeID
    awayTeamId: NodeID
}

export type Change = Operation

export interface PropertyChange {
    sourceValue: unknown
    change: Change
}

export interface Discrepancy {
    subjectId: NodeID
    subjectType: DiscrepancySubject
    parentId: NodeID | null
    eventMeta: EventMeta
    propertyChange: PropertyChange
}

export type Filters = Partial<Pick<Discrepancy, "subjectId" | "subjectType">>

export interface DiscrepanciesRepository {
    discrepancies: (filters: Filters) => Promise<Discrepancy[]>
    insert: (discrepancy: Discrepancy) => Promise<void>
}

export class DiscrepanciesRepositoryMemory implements DiscrepanciesRepository {
    private discrepanciesStore: Discrepancy[] = []

    discrepancies (filters?: Filters): Promise<Discrepancy[]> {
        if (filters) {
            validateFilters(filters)
            return Promise.resolve(this.discrepanciesStore.filter(this.apply(filters)))
        }
        return Promise.resolve([...this.discrepanciesStore])
    }

    insert (discrepancy: Discrepancy): Promise<void> {
        this.discrepanciesStore.push(discrepancy)
        return Promise.resolve(undefined)
    }

    private apply = (filters: Filters) => (discrepancy: Discrepancy): boolean => {
        const { subjectId, subjectType } = filters

        if (subjectId && discrepancy.subjectId !== subjectId) {
            return false
        }

        if (subjectType && discrepancy.subjectType !== subjectType) {
            return false
        }

        return true
    }
}

const validateFilters = (filters: Filters): void => {
    if (filters) {
        const { subjectType } = filters
        if (subjectType && !ALL_DISCREPANCY_SUBJECTS.includes(subjectType)) {
            throw new InvalidFilterError('subjectType', subjectType, ALL_DISCREPANCY_SUBJECTS.join(' | '))
        }
    }
}
