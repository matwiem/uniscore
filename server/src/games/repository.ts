export interface GamesRepository {
    game: (id: string) => Promise<Game | null>
    games: () => Promise<Game[]>
    insert: (game: Game) => Promise<void>
}

export class Game {
    id: string
    statistics: GameStatistics
    home: Team
    away: Team

    constructor(id: string, statistics: GameStatistics, home: Team, away: Team) {
        this.id = id
        this.statistics = statistics
        this.home = home
        this.away = away
    }
}

export interface GameStatistics {
    attendance: number
}

export class Team {
    id: string
    statistics: TeamStatistics
    players: {[key: string]: Player}

    constructor(id: string, statistics: TeamStatistics, players: {[key: string]: Player}) {
        this.id = id
        this.statistics = statistics
        this.players = players
    }
}

export interface TeamStatistics {
    rushAttempts: number
    rushTds: number
    rushYdsGained: number
    rec: number
    recYards: number
}

export class Player {
    id: string
    statistics: PlayerStatistics

    constructor(id: string, statistics: PlayerStatistics) {
        this.id = id
        this.statistics = statistics
    }
}

export interface PlayerStatistics {
    rushAttempts: number | null
    rushTds: number | null
    rushYdsGained: number | null
    rec: number | null
    recYards: number | null
}

export class GamesRepositoryMemory implements GamesRepository {
    gamesStore: {[key: string]: Game} = {}

    game (id: string): Promise<Game | null> {
        return Promise.resolve(this.gamesStore[id] || null)
    }

    games (): Promise<Game[]> {
        return Promise.resolve(Object.values(this.gamesStore))
    }

    insert (game: Game): Promise<void> {
        this.gamesStore[game.id] = game
        return Promise.resolve(undefined)
    }
}
