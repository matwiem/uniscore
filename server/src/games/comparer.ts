import { compare, getValueByPointer } from 'fast-json-patch'

import { Game, Player, Team } from '@src/games/repository'
import {
    Discrepancy,
    DiscrepancySubject, EventMeta,
    NodeID,
} from '@src/discrepancies/repository'
import { InvalidStateError } from '@src/games/errors'

export interface Comparer {
    compareGame: (game: Game) => void
    compareTeam: (team: Team) => void
    comparePlayer: (player: Player) => void
}

export interface Comparable {
    accept: (comparer: Comparer) => void
}

export class ComparerDemo implements Comparer {
    private sourceGame: Game | null = null
    private sourceTeam: Team | null = null
    private sourcePlayer: Player | null = null
    private discrepancies: Discrepancy[] = []

    getDiscrepancies(): Discrepancy[] {
        return this.discrepancies
    }

    setSourceGame(game: Game) {
        this.reset()
        this.sourceGame = game
    }

    compareGame(target: Game) {
        if (this.sourceGame === null) {
            throw new InvalidStateError("Cannot compare entities, sourceGame is null")
        }

        this.discrepancies = [...this.discrepancies, ...this.compare(
            this.sourceGame.id,
            "GAME",
            null,
            this.eventMeta(),
            {statistics: this.sourceGame.statistics},
            {statistics: target.statistics},
        )]

        this.sourceTeam = this.sourceGame.home
        target.home.accept(this)
        this.sourceTeam = this.sourceGame.away
        target.away.accept(this)
    }

    compareTeam(target: Team) {
        if (this.sourceGame === null) {
            throw new InvalidStateError("Cannot compare entities, sourceGame is null")
        }

        if (this.sourceTeam === null) {
            throw new InvalidStateError("Cannot compare entities, sourceTeam is null")
        }

        this.discrepancies = [...this.discrepancies, ...this.compare(
            this.sourceTeam.id,
            "TEAM",
            this.sourceGame.id,
            this.eventMeta(),
            {statistics: this.sourceTeam.statistics},
            {statistics: target.statistics},
        )]

        for (const player of Object.values(this.sourceTeam.players)) {
            this.sourcePlayer = player
            if (target.players[this.sourcePlayer.id]) {
                target.players[this.sourcePlayer.id].accept(this)
            }
        }
    }

    comparePlayer(target: Player) {
        if (this.sourceGame === null) {
            throw new InvalidStateError("Cannot compare entities, sourceGame is null")
        }

        if (this.sourceTeam === null) {
            throw new InvalidStateError("Cannot compare entities, sourceTeam is null")
        }

        if (this.sourcePlayer === null) {
            throw new InvalidStateError("Cannot compare entities, sourcePlayer is null")
        }

        this.discrepancies = [...this.discrepancies, ...this.compare(
            this.sourcePlayer.id,
            "PLAYER",
            this.sourceTeam.id,
            this.eventMeta(),
            {statistics: this.sourcePlayer.statistics},
            {statistics: target.statistics}
        )]
    }

    private compare(subjectId: NodeID, subjectType: DiscrepancySubject, parentId: NodeID | null, eventMeta: EventMeta, c1: ComparableProperties, c2: ComparableProperties): Discrepancy[] {
        const operations = compare(c1, c2)
        return operations.map(op => ({
            subjectId,
            subjectType,
            parentId,
            eventMeta,
            propertyChange: {
                sourceValue: getValueByPointer(c1, op.path) as unknown,
                change: op,
            },
        }))
    }

    private eventMeta(): EventMeta {
        if (this.sourceGame === null) {
            throw new InvalidStateError("Cannot compare entities, sourceGame is null")
        }
        return {
            gameId: this.sourceGame.id,
            homeTeamId: this.sourceGame.home.id,
            awayTeamId: this.sourceGame.away.id,
        }
    }

    private reset(): void {
        this.sourceGame = null
        this.sourceTeam = null
        this.sourceGame = null
        this.discrepancies = []
    }
}

type ComparableProperties =
    | Omit<Game, "id" | "sourceId" | "home" | "away" | "accept">
    | Omit<Team, "id" | "players" | "accept">
    | Omit<Player, "id" | "accept">
