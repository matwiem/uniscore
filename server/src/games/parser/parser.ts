import { Game } from '@src/games/repository/repository'

export interface Parser {
    parse: (data: unknown) => Game
}
