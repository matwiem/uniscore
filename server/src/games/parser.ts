import { Game } from '@src/games/repository'

export interface Parser {
    parse: (data: unknown) => Game
}
