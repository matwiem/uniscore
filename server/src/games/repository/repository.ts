import { GameExistsError } from '@src/games/errors'
import { Comparable, Comparer } from '@src/games/comparer/comparer'

export interface GamesRepository {
    game: (id: string) => Promise<Game | null>
    games: () => Promise<Game[]>
    insert: (game: Game) => Promise<void>
    insertMany: (games: Game[]) => Promise<void>
}

export class Game implements Comparable{
    id: string
    sourceId: string
    statistics: GameStatistics
    home: Team
    away: Team

    constructor(id: string, sourceId: string, statistics: GameStatistics, home: Team, away: Team) {
        this.id = id
        this.sourceId = sourceId
        this.statistics = statistics
        this.home = home
        this.away = away
    }

    accept (comparer: Comparer): void {
        comparer.compareGame(this)
    }
}

export interface GameStatistics {
    attendance: number
}

export class Team implements Comparable {
    id: string
    statistics: TeamStatistics
    players: {[key: string]: Player}

    constructor(id: string, statistics: TeamStatistics, players: {[key: string]: Player}) {
        this.id = id
        this.statistics = statistics
        this.players = players
    }

    accept (comparer: Comparer): void {
        comparer.compareTeam(this)
    }
}

export interface TeamStatistics {
    rushAttempts: number
    rushTds: number
    rushYdsGained: number
    rec: number
    recYards: number
}

export class Player implements Comparable {
    id: string
    statistics: PlayerStatistics

    constructor(id: string, statistics: PlayerStatistics) {
        this.id = id
        this.statistics = statistics
    }

    accept (comparer: Comparer): void {
        comparer.comparePlayer(this)
    }
}

export interface PlayerStatistics {
    rushAttempts: number | null
    rushTds: number | null
    rushYdsGained: number | null
    rec: number | null
    recYards: number | null
}

export class GamesRepositoryMemory implements GamesRepository {
    private gamesStore: {[key: string]: Game} = {}

    game (id: string): Promise<Game | null> {
        return Promise.resolve(this.gamesStore[id] || null)
    }

    games (): Promise<Game[]> {
        return Promise.resolve(Object.values(this.gamesStore))
    }

    insert (game: Game): Promise<void> {
        if (this.gamesStore[game.id]) {
            throw new GameExistsError(game.id)
        }
        this.gamesStore[game.id] = game
        return Promise.resolve(undefined)
    }

    async insertMany (games: Game[]): Promise<void> {
        for (const game of games) {
            await this.insert(game)
        }
        return Promise.resolve(undefined)
    }
}
