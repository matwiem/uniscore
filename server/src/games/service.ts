import jsonfile from 'jsonfile'

import { GamesRepository } from '@src/games/repository'
import { Parser } from '@src/games/parser'

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface GamesService {}

export class GamesServiceDemo implements GamesService {
    private sourceGamesRepository: GamesRepository
    private targetGamesRepository: GamesRepository

    constructor (sourceGamesRepository: GamesRepository, targetGamesRepository: GamesRepository) {
        this.sourceGamesRepository = sourceGamesRepository
        this.targetGamesRepository = targetGamesRepository
    }

    loadSourceFromFile = async (path: string, parser: Parser) => {
        await this.loadGamesFromFile(path, this.sourceGamesRepository, parser)
    }

    loadTargetFromFile = async (path: string, parser: Parser) => {
        await this.loadGamesFromFile(path, this.targetGamesRepository, parser)
    }

    loadSourceFromCode = async (data: unknown, parser: Parser) => {
        await this.loadGamesFromCode(data, this.sourceGamesRepository, parser)
    }

    loadTargetFromCode = async (data: unknown, parser: Parser) => {
        await this.loadGamesFromCode(data, this.targetGamesRepository, parser)
    }

    private loadGamesFromFile = async (path: string, repository: GamesRepository, parser: Parser): Promise<void> => {
        const data: unknown = await jsonfile.readFile(path)
        await this.loadGamesFromCode(data, repository, parser)
    }

    private loadGamesFromCode = async (data: unknown, repository: GamesRepository, parser: Parser): Promise<void> => {
        if (Array.isArray(data)) {
            await repository.insertMany(data.map(parser.parse))
        } else {
            await repository.insert(parser.parse(data))
        }
    }
}
