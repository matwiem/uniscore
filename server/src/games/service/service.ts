import jsonfile from 'jsonfile'

import { GamesRepository } from '@src/games/repository/repository'
import { Parser } from '@src/games/parser/parser'
import {
    DiscrepanciesRepository
} from '@src/discrepancies/repository/repository'
import { Comparer } from '@src/games/comparer/comparer'

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface GamesService {}

export class GamesServiceDemo implements GamesService {
    private sourceGamesRepository: GamesRepository
    private targetGamesRepository: GamesRepository
    private discrepanciesRepository: DiscrepanciesRepository

    constructor (sourceGamesRepository: GamesRepository, targetGamesRepository: GamesRepository, discrepanciesRepository: DiscrepanciesRepository) {
        this.sourceGamesRepository = sourceGamesRepository
        this.targetGamesRepository = targetGamesRepository
        this.discrepanciesRepository = discrepanciesRepository
    }

    compareGames = async (comparer: Comparer): Promise<void> => {
        const sourceGames = await this.sourceGamesRepository.games()
        for (const sourceGame of sourceGames) {
            comparer.setSourceGame(sourceGame)
            const targetGame = await this.targetGamesRepository.game(sourceGame.id)
            if (targetGame) {
                targetGame.accept(comparer)
            }
            const discrepancies = comparer.getDiscrepancies()
            await this.discrepanciesRepository.insertMany(discrepancies)
        }
    }

    loadSourceFromFile = async (path: string, parser: Parser) => {
        await this.loadGamesFromFile(path, this.sourceGamesRepository, parser)
    }

    loadTargetFromFile = async (path: string, parser: Parser) => {
        await this.loadGamesFromFile(path, this.targetGamesRepository, parser)
    }

    loadSourceFromCode = async (data: unknown, parser: Parser) => {
        await this.loadGamesFromCode(data, this.sourceGamesRepository, parser)
    }

    loadTargetFromCode = async (data: unknown, parser: Parser) => {
        await this.loadGamesFromCode(data, this.targetGamesRepository, parser)
    }

    private loadGamesFromFile = async (path: string, repository: GamesRepository, parser: Parser): Promise<void> => {
        const data: unknown = await jsonfile.readFile(path)
        await this.loadGamesFromCode(data, repository, parser)
    }

    private loadGamesFromCode = async (data: unknown, repository: GamesRepository, parser: Parser): Promise<void> => {
        if (Array.isArray(data)) {
            await repository.insertMany(data.map(parser.parse))
        } else {
            await repository.insert(parser.parse(data))
        }
    }
}
