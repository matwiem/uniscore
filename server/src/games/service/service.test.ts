import { GamesRepositoryMemory } from '../repository/repository'
import { GamesServiceDemo } from './service'
import { ParserSR } from '../parser/parser.sr'
import { ParserExternal } from '../parser/parser.external'
import { gameDtoSr } from '../parser/parser.sr.mock'
import { gameDtoExternal } from '../parser/parser.external.mock'
import { ComparerDemo } from '../comparer/comparer'
import {
    DiscrepanciesRepositoryMemory
} from '../../discrepancies/repository/repository'

describe(`GamesServiceDemo`, () => {
    it(`should parse game DTOs and store them in a repository`, async () => {
        const repositorySr = new GamesRepositoryMemory()
        const repositoryExternal = new GamesRepositoryMemory()
        const discrepancyRepository = new DiscrepanciesRepositoryMemory()
        const gamesService = new GamesServiceDemo(repositorySr, repositoryExternal, discrepancyRepository)
        const parserSr = new ParserSR()
        const parserExternal = new ParserExternal()
        await gamesService.loadSourceFromCode(gameDtoSr, parserSr)
        await gamesService.loadTargetFromCode(gameDtoExternal, parserExternal)

        const gamesSr = await repositorySr.games()
        expect(gamesSr)
            .toHaveLength(1)
        expect(gamesSr)
            .toEqual(expect.arrayContaining([expect.objectContaining({id: gameDtoSr.game.id})]))

        const gamesExternal = await repositoryExternal.games()
        expect(gamesExternal)
            .toHaveLength(1)
        expect(gamesExternal)
            .toEqual(expect.arrayContaining([expect.objectContaining({id: gameDtoExternal.game.id})]))
    })

    it(`should find discrepancies between source and target games repositories and insert them into discrepancies repository`, async () => {
        const repositorySr = new GamesRepositoryMemory()
        const repositoryExternal = new GamesRepositoryMemory()
        const discrepanciesRepository = new DiscrepanciesRepositoryMemory()
        const gamesService = new GamesServiceDemo(repositorySr, repositoryExternal, discrepanciesRepository)
        const parserSr = new ParserSR()
        const parserExternal = new ParserExternal()
        await gamesService.loadSourceFromCode(gameDtoSr, parserSr)
        await gamesService.loadTargetFromCode(gameDtoExternal, parserExternal)
        const comparer = new ComparerDemo()
        await gamesService.compareGames(comparer)
        const discrepancies = await discrepanciesRepository.discrepancies()
        expect(discrepancies)
            .toHaveLength(15)
        expect(discrepancies)
            .toMatchSnapshot()
    })
})
