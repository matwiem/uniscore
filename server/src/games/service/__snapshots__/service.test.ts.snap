// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamesServiceDemo should find discrepancies between source and target games repositories and insert them into discrepancies repository 1`] = `
[
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": null,
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/attendance",
        "value": 33876,
      },
      "sourceValue": 62487,
    },
    "subjectId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "subjectType": "GAME",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rec",
        "value": 31,
      },
      "sourceValue": 257,
    },
    "subjectId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "subjectType": "TEAM",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/recYards",
        "value": 257,
      },
      "sourceValue": 31,
    },
    "subjectId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "subjectType": "TEAM",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rushYdsGained",
        "value": 90,
      },
      "sourceValue": 78,
    },
    "subjectId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "subjectType": "TEAM",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rec",
        "value": 1,
      },
      "sourceValue": 8,
    },
    "subjectId": "e5b67a2e-e0f7-4646-a343-c9bd0ae447f0",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/recYards",
        "value": 4,
      },
      "sourceValue": 41,
    },
    "subjectId": "e5b67a2e-e0f7-4646-a343-c9bd0ae447f0",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rushYdsGained",
        "value": -15,
      },
      "sourceValue": -9,
    },
    "subjectId": "c90711f2-5c4e-4f2e-8635-0d3f2336f581",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/recYards",
        "value": 92,
      },
      "sourceValue": 88,
    },
    "subjectId": "713daff7-86cc-4283-91a8-1579c3f58f59",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rec",
        "value": 27,
      },
      "sourceValue": 239,
    },
    "subjectId": "ba27615c-07df-41f0-864b-332575f744f2",
    "subjectType": "TEAM",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/recYards",
        "value": 230,
      },
      "sourceValue": 27,
    },
    "subjectId": "ba27615c-07df-41f0-864b-332575f744f2",
    "subjectType": "TEAM",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rushYdsGained",
        "value": 105,
      },
      "sourceValue": 97,
    },
    "subjectId": "ba27615c-07df-41f0-864b-332575f744f2",
    "subjectType": "TEAM",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "ba27615c-07df-41f0-864b-332575f744f2",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rushYdsGained",
        "value": 73,
      },
      "sourceValue": 76,
    },
    "subjectId": "e11a2815-8f31-44b0-8c37-64240bcbc059",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "ba27615c-07df-41f0-864b-332575f744f2",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rushTds",
        "value": 2,
      },
      "sourceValue": 1,
    },
    "subjectId": "e11a2815-8f31-44b0-8c37-64240bcbc059",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "ba27615c-07df-41f0-864b-332575f744f2",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/recYards",
        "value": 47,
      },
      "sourceValue": 46,
    },
    "subjectId": "f94fb8a3-cd77-4064-bc65-447550b43622",
    "subjectType": "PLAYER",
  },
  {
    "eventMeta": {
      "awayTeamId": "ba27615c-07df-41f0-864b-332575f744f2",
      "gameId": "15adf794-5630-4dc4-b0e7-f8d437b585b1",
      "homeTeamId": "344e408b-846e-44ac-b491-7802ab0f7af0",
    },
    "parentId": "ba27615c-07df-41f0-864b-332575f744f2",
    "propertyChange": {
      "change": {
        "op": "replace",
        "path": "/statistics/rec",
        "value": 9,
      },
      "sourceValue": 8,
    },
    "subjectId": "f94fb8a3-cd77-4064-bc65-447550b43622",
    "subjectType": "PLAYER",
  },
]
`;
